<launch>
<!-- SCENE : Marina Bay Sands -->
<arg name="fov_h" value="55.0"/>
<arg name="fov_w" value="75.0"/>
<arg name="cx" value="320.0"/>
<arg name="cy" value="240.0"/>
<arg name="fx" value="417.0467"/>
<arg name="fy" value="461.0951"/>
<arg name="max_dist" value="11.0"/>
<arg name="resolution_" value="0.4"/>
<arg name="max_vel" value="2.0"/>
<arg name="max_acc" value="1.0"/>
<arg name="max_jerk" value="0.5"/>
<arg name="max_yd" value="60.0"/>
<arg name="max_ydd" value="30.0"/>
<arg name="drone_radius" value="0.3"/>

<node pkg ="hierarchical_coverage_planner" name ="hcplanner_exec" type ="hcplanner_exec" output = "screen" required ="true">
  <!-- ! Skeleton-based Space Decomposition ! -->
  <param name="rosa_main/estimation_num" value="2000" type="int"/>
  <param name="rosa_main/pcd" value="$(find hierarchical_coverage_planner)/data/MBS.pcd" type="string"/>
  <param name="rosa_main/radius" value="0.1" type="double"/>
  <param name="rosa_main/th_mah" value="0.01" type="double"/>
  <param name="rosa_main/delta" value="0.01" type="double"/>
  <param name="rosa_main/num_drosa" value="5" type="int"/>
  <param name="rosa_main/num_dcrosa" value="2" type="int"/>
  <param name="rosa_main/k_KNN" value="6" type="int"/>
  <param name="rosa_main/sample_r" value="0.05" type="double"/>
  <param name="rosa_main/alpha" value="0.3" type="double"/>
  <param name="rosa_main/pt_downsample_size" value="0.02" type="double"/>
  <param name="rosa_main/estimation_number" value="10" type="int"/>
  <param name="rosa_main/upper_bound_angle_inner_decomp" value="45.0" type="double"/>
  <param name="rosa_main/upper_bound_length_inner_decomp" value="1.0" type="double"/>
  <param name="rosa_main/Prune" value="false" type="bool"/>
  <param name="rosa_main/lower_bound_length" value="0.1" type="double"/>
  <param name="rosa_main/lower_bound_prune_angle" value="75.0" type="double"/>
  <param name="rosa_main/upper_bound_original_points_num" value="10000" type="int"/>
  <param name="rosa_main/Ground" value="true" type="bool"/>

  <!-- ! Mapping ! -->
  <param name="hcmap/resolution" value="$(arg resolution_)" type="double"/>
  <param name="hcmap/interval" value="1.0" type="double"/>
  <param name="hcmap/plane_thickness" value="3.0" type="double"/>
  <param name="hcmap/dilateRadius" value="0.5" type="double"/>
  <param name="hcmap/checkScale" value="5.0" type="double"/>
  <param name="hcmap/checkSize" value="7" type="int"/>
  <param name="hcmap/inflateVoxel" value="4" type="int"/>

  <!-- ! Perception ! -->
  <param name="perception_utils/top_angle" value="$(eval arg('fov_h')*3.1415926/360.0)" type="double"/>
  <param name="perception_utils/left_angle" value="$(eval arg('fov_w')*3.1415926/360.0)" type="double"/>
  <param name="perception_utils/right_angle" value="$(eval arg('fov_w')*3.1415926/360.0)" type="double"/>
  <param name="perception_utils/max_dist" value="$(arg max_dist)" type="double"/>
  <param name="perception_utils/vis_dist" value="3.0" type="double"/>

  <!-- ! Path Searching ! -->
  <param name="astar/lambda_heu_hc" value="10.0" type="double"/>
  <param name="astar/resolution_astar" value="$(arg resolution_)" type="double"/>
  <param name="astar/allocate_num" value="1000000" type="int"/>
  <param name="astar/max_search_time" value="0.0075" type="double"/>
  <param name="astar/searching_res" value="0.4" type="double"/>

  <!-- ! Hierarchical Coverage Planner ! -->
  <param name="hcplanner/mesh" value="$(find hierarchical_coverage_planner)/data/mesh/MBS.obj" type="string"/>
  <param name="hcplanner/fullcloud" value="$(find hierarchical_coverage_planner)/data/MBSmore.pcd" type="string"/>
  <param name="hcplanner/model_downsample_size" value="2.0" type="double"/>
  <param name="hcplanner/safe_inner_dist_coeff" value="1.1" type="double"/>
  <param name="hcplanner/cx" value="$(arg cx)" type="double"/>
  <param name="hcplanner/cy" value="$(arg cy)" type="double"/>
  <param name="hcplanner/fx" value="$(arg fx)" type="double"/>
  <param name="hcplanner/fy" value="$(arg fy)" type="double"/>
  <param name="hcplanner/tripod_head" value="false" type="bool"/>
  <param name="hcplanner/sample_step4normal" value="1.0" type="double"/>
  <param name="hcplanner/current_x_" value="0.0" type="double"/>
  <param name="hcplanner/current_y_" value="20.0" type="double"/>
  <param name="hcplanner/current_z_" value="-13.0" type="double"/>
  <param name="hcplanner/exec_mode" value="release" type="string"/> <!-- mode : debug or release -->
  
  <!-- ! Viewpoint Manager ! -->
  <param name="viewpoint_manager/visible_range" value="$(arg max_dist)" type="double"/>
  <param name="viewpoint_manager/viewpoints_distance" value="8.0" type="double"/>
  <param name="viewpoint_manager/fov_h" value="$(arg fov_h)" type="double"/>
  <param name="viewpoint_manager/fov_w" value="$(arg fov_w)" type="double"/>
  <param name="viewpoint_manager/pitch_upper" value="70.0" type="double"/>
  <param name="viewpoint_manager/pitch_lower" value="-85.0" type="double"/>
  <param name="viewpoint_manager/zGround" value="true" type="bool"/>
  <param name="viewpoint_manager/GroundPos" value="-10.5" type="double"/>
  <param name="viewpoint_manager/safeHeight" value="1.5" type="double"/>
  <param name="viewpoint_manager/safe_radius" value="3.0" type="double"/>
  <param name="viewpoint_manager/attitude_type" value="all" type="string"/> <!-- attitude_type : 1) yaw 2) all -> (pitch, yaw) -->
  <param name="viewpoint_manager/max_iter_num" value="3" type="int"/> 
  <param name="viewpoint_manager/pose_update" value="true" type="bool"/> 

  <!-- ! Path Solver ! -->
  <param name="hcplanner/precision_" value="100" type="int"/>
  <param name="hcplanner/global_solver" value="$(find hierarchical_coverage_planner)/solution/LKH" type="string"/>
  <param name="hcplanner/global_par_file" value="$(find hierarchical_coverage_planner)/solution/global/global.par" type="string"/>
  <param name="hcplanner/global_problem_file" value="$(find hierarchical_coverage_planner)/solution/global/global.tsp" type="string"/>
  <param name="hcplanner/global_result_file" value="$(find hierarchical_coverage_planner)/solution/global/global.txt" type="string"/>
  <param name="hcplanner/global_runs" value="1" type="int"/>
  <param name="hcplanner/local_folder" value="$(find hierarchical_coverage_planner)/solution/local" type="string"/>
  <param name="hcplanner/local_runs" value="1" type="int"/>
  <param name="hcplanner/vmax_" value="$(arg max_vel)" type="double"/>
  <param name="hcplanner/amax_" value="$(arg max_acc)" type="double"/>
  <param name="hcplanner/jmax_" value="$(arg max_jerk)" type="double"/>
  <param name="hcplanner/ydmax_" value="$(eval arg('max_yd')*3.1415926/180.0)" type="double"/>
  <param name="hcplanner/yddmax_" value="$(eval arg('max_ydd')*3.1415926/180.0)" type="double"/>
  <param name="hcplanner/JointRadius" value="1.5" type="double"/>
  <param name="hcplanner/local2opt_trial" value="5000" type="int"/>

  <!-- ! Trajectory Generation ! -->
  <param name="hctraj/drone_radius" value="$(arg drone_radius)" type="double"/>
  <param name="hctraj/vmax_" value="$(arg max_vel)" type="double"/>
  <param name="hctraj/amax_" value="$(arg max_acc)" type="double"/>
  <param name="hctraj/jmax_" value="$(arg max_jerk)" type="double"/>
  <param name="hctraj/ydmax_" value="$(eval arg('max_yd')*3.1415926/180.0)" type="double"/>
  <param name="hctraj/rho_T" value="500.0" type="double"/>
  <param name="hctraj/rho_j" value="800.0" type="double"/>
  <param name="hctraj/rho_a" value="800.0" type="double"/>
  <param name="hctraj/rho_v" value="800.0" type="double"/>
  <param name="hctraj/rho_p" value="800.0" type="double"/>
  <param name="hctraj/rho_z" value="2000.0" type="double"/>
  <param name="hctraj/rho_e" value="500.0" type="double"/>
  <param name="hctraj/TrajFile" value="$(find hierarchical_coverage_planner)/solution/Traj/TrajInfoMBS.txt" type="string"/>
  <param name="hctraj/CloudFile" value="$(find hierarchical_coverage_planner)/solution/Traj/CloudInfoMBS.txt" type="string"/>
  <param name="hctraj/PosFile" value="$(find hierarchical_coverage_planner)/solution/Traj/PositionCoeffMBS.txt" type="string"/>
  <param name="hctraj/PitchFile" value="$(find hierarchical_coverage_planner)/solution/Traj/PitchCoeffMBS.txt" type="string"/>
  <param name="hctraj/YawFile" value="$(find hierarchical_coverage_planner)/solution/Traj/YawCoeffMBS.txt" type="string"/>

  <!-- ! FC-Planner ! -->
  <param name="hcopp/TimeStamp" value="0.3" type="double"/>
  <param name="hcopp/TrajSampleFrequency" value="0.1" type="double"/>
  <param name="hcopp/droneMesh" value="$(find traj_utils)/f250.dae" type="string"/>
</node>

</launch>